mmengine/testing/runner_test_case.py

```diff
@@ -0,0 +1,172 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import copy
+import logging
+import os
+import tempfile
+import time
+from unittest import TestCase
+from uuid import uuid4
+
+import torch
+import torch.nn as nn
+from torch.distributed import destroy_process_group
+from torch.utils.data import Dataset
+
+import mmengine.hooks  # noqa F401
+import mmengine.optim  # noqa F401
+from mmengine.config import Config
+from mmengine.dist import is_distributed
+from mmengine.evaluator import BaseMetric
+from mmengine.logging import MessageHub, MMLogger
+from mmengine.model import BaseModel
+from mmengine.registry import DATASETS, METRICS, MODELS, DefaultScope
+from mmengine.runner import Runner
+from mmengine.visualization import Visualizer
+
+
+class ToyModel(BaseModel):
+
+    def __init__(self, data_preprocessor=None):
+        super().__init__(data_preprocessor=data_preprocessor)
+        self.linear1 = nn.Linear(2, 2)
+        self.linear2 = nn.Linear(2, 1)
+
+    def forward(self, inputs, data_samples, mode='tensor'):
+        if isinstance(inputs, list):
+            inputs = torch.stack(inputs)
+        if isinstance(data_samples, list):
+            data_sample = torch.stack(data_samples)
+        outputs = self.linear1(inputs)
+        outputs = self.linear2(outputs)
+
+        if mode == 'tensor':
+            return outputs
+        elif mode == 'loss':
+            loss = (data_sample - outputs).sum()
+            outputs = dict(loss=loss)
+            return outputs
+        elif mode == 'predict':
+            return outputs
+
+
+class ToyDataset(Dataset):
+    METAINFO = dict()  # type: ignore
+    data = torch.randn(12, 2)
+    label = torch.ones(12)
+
+    @property
+    def metainfo(self):
+        return self.METAINFO
+
+    def __len__(self):
+        return self.data.size(0)
+
+    def __getitem__(self, index):
+        return dict(inputs=self.data[index], data_samples=self.label[index])
+
+
+class ToyMetric(BaseMetric):
+
+    def __init__(self, collect_device='cpu', dummy_metrics=None):
+        super().__init__(collect_device=collect_device)
+        self.dummy_metrics = dummy_metrics
+
+    def process(self, data_batch, predictions):
+        result = {'acc': 1}
+        self.results.append(result)
+
+    def compute_metrics(self, results):
+        return dict(acc=1)
+
+
+class RunnerTestCase(TestCase):
```

Comment:
I think it's better to add some docstring here, which may help understand what it's designed for and how to use it.